### ISSUE TITLE
Inherent async fn returning `Self` treats type's lifetime parameters as `'static`

### ISSUE THREAD

**Issue by Arnavion** [rust-lang/rust#61949]
`rustc 1.37.0-nightly (2887008e0 2019-06-12)` and playground's `2019-06-17 b25ee644971a168287ee`

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=50cf396172d221ccf89b61e7c4719fbd)

```rust
#![feature(async_await)]

pub struct Foo<'a> {
    pub bar: &'a i32,
}

impl<'a> Foo<'a> {
    pub async fn new(bar: &'a i32) -> Self {
        Foo {
            bar
        }
    }
}
```

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
 --> src/lib.rs:8:22
  |
8 |     pub async fn new(bar: &'a i32) -> Self {
  |                      ^^^
  |
note: first, the lifetime cannot outlive the lifetime 'a as defined on the impl at 7:6...
 --> src/lib.rs:7:6
  |
7 | impl<'a> Foo<'a> {
  |      ^^
  = note: ...so that the expression is assignable:
          expected &i32
             found &'a i32
  = note: but, the lifetime must be valid for the static lifetime...
  = note: ...so that the types are compatible:
          expected Foo<'_>
             found Foo<'static>
```

---

It works to either change the signature to take `bar: &'static i32`, or to change the body of the fn to use a static borrow like `bar: &5`. So the compiler really does want the function to return a `Foo<'static>`, even though `Self` is a `Foo<'a>`

---

The workaround is to not use `Self`:

```diff
-    pub async fn new(bar: &'a i32) -> Self {
+    pub async fn new(bar: &'a i32) -> Foo<'a> {
```

---

Mentoring notes: See [notes here on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/.60Self.60.20and.20.60'static.60.20.2361949/near/170476808).

**Comment by Arnavion** [comment]
jebrosen pointed out on IRC that it also happens with RPIT in general, so this might be the same as https://github.com/rust-lang/rust/issues/53613

**Comment by nikomatsakis** [comment]
As I pointed out in #53613, there is a more dangerous variant of this involving projections:

```rust
#![feature(async_await)]

pub trait HasItem<'a> {
    type Item;
}

// Does not compile:
async fn example1<'a, T: HasItem<'a>>(item: T::Item) -> T::Item {
    item
}


fn main() { }
```

**Comment by nikomatsakis** [comment]
There are some forward compatibility concerns here. In particular, callers can (somewhat incorrectly) assume that the returned futures are `'static`. In other words, the callee is committing to returning a future that does not capture `'a` in these examples -- right now, that typically means `callee` doesn't compile, but it could happen that the callee *happens* to compile, and then callers would be able to use this future in places where `'a` is out of scope. If we later fix this bug, those callers will stop compiling.

**Comment by nikomatsakis** [comment]
The bug involving `Self` is rather straightforward (but grungy) to fix. It's also possible for us to, temporarily, just create an error if you return a type mentioning `Self` in an `async fn`.

But fixing the `T::Item` case is rather more complex. We don't know *until type-checking* that `T::Item` resolves to `<T as HasItem<'a>>::Item`. And, if I had my druthers, we might not know until *even later than that*. I suppose we could do a forward compatibility restriction of some kind.

**Comment by nikomatsakis** [comment]
An example of something which compiles now but I think should not (and would not, if this were fixed):

```rust
#![feature(async_await)]

pub struct Foo<'a> {
    pub bar: &'a i32,
}

impl<'a> Foo<'a> {
    pub async fn new(_bar: &'a i32) -> Self {
        Foo {
            bar: &22
        }
    }
}

async fn foo() {
  let x = {
    let bar = 22;
    Foo::new(&bar).await
  };
  drop(x);
}

fn main() { }
```

the problem here is that the type of `x` should be `Foo<'bar>` -- i.e., it should be constrained to the lifetime of `bar`, but it is actually `Foo<'static>`.

**Comment by nikomatsakis** [comment]
I left some notes on how to fix this on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/.60Self.60.20and.20.60'static.60.20.2361949/near/170476808)

**Comment by nikomatsakis** [comment]
OK so having thought this over I have come to the conclusion that the *right* behavior here is to give an error if people use `Self` or `T::Foo` projections from within an `impl Trait` such that they wind up "inheriting lifetimes". It's not an ideal solution but it'll permit many uses while forbidding the cases that are causing trouble and giving us time to fix properly.

The way I think we should do this:

During type checking, there is a [`check_opaque`](https://github.com/rust-lang/rust/blob/b8ec4c4d11ede0fba333a0474ed473dbe82aacf1/src/librustc_typeck/check/mod.rs#L1317-L1331) function that currently checks for cyclic opaque types.

In the case of existential types whose [origin](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/struct.ExistTy.html#structfield.origin) is either [`ReturnImplTrait`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/enum.ExistTyOrigin.html#variant.ReturnImplTrait) or [`AsyncFn`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/hir/enum.ExistTyOrigin.html#variant.AsyncFn), we want to enforce this condition.

What we would do is to access the `tcx.predicates_of` the existential def-id. The [`predicates`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.GenericPredicates.html#structfield.predicates) field of the result will contain the "bounds" of the impl trait (with the self-type being an opaque type). (Note that we don't want to inspect the [parent](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.GenericPredicates.html#structfield.parent), which will contain predicates defined in the enclosing scope.)

Actually, we *may* want to introduce a new query (`existential_bounds_of` or something) that just returns those predicates that `tcx.predicates_of` would return..? I guess it's fine to invoke `tcx.predicates_of`, it just feels a touch fragile. Ah well, that's what tests are for. =)

Anyway, we basically iterate over that `predicates` field using a [type visitor](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/fold/trait.TypeVisitor.html). In there, we look for references to regions that are [`ReEarlyBound`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/enum.RegionKind.html#variant.ReEarlyBound) but where the [`index`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.EarlyBoundRegion.html#structfield.index) indicates it is coming from the parent. We would report an error if we found any such region.

One thing to watch out for though while we are visiting the type. If we have something like `impl Foo`, that will be represented by a predicate `O: Foo` where `O` is the *opaque identity type*. Such a type is constructed here:

https://github.com/rust-lang/rust/blob/b8ec4c4d11ede0fba333a0474ed473dbe82aacf1/src/librustc_typeck/collect.rs#L1924-L1925

Such a type will contain the illegal regions and we need to ignore it. We basically want to ignore instances of the self-type when visiting -- i.e., we can construct the opaque identity type (call it `opaque_identity_ty`) and override `visit_ty` to skip visiting the contents if `ty == opaque_identity_ty`.

**Comment by Centril** [comment]
From Lang team meeting: @cramertj will do a more targeted fix for `async fn` based on https://github.com/rust-lang/rust/pull/62849. If crater completes in time then we can generalize #62849.

**Comment by Centril** [comment]
Update: https://github.com/rust-lang/rust/pull/62849 has landed. The `-Blocking` aspect of this has been fixed so I am relabeling as `-Deferred`.

**Comment by Arnavion** [comment]
Did the new error in #62849 not make it to stable 1.39.0 ? I still see the old error about lifetime being inferred as `'static` and not matching, rather than the explicit "`async fn` return type cannot contain a projection or `Self`..." error added by that PR.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=06a794508330334476e5b9f298ea233b

**Comment by SergioBenitez** [comment]
I'm _very_ surprised that this is currently shipping in stable due to the forwards/backward compatibility hazards mentioned by @nikomatsakis. The error message ostensibly introduced by #62849 _is not_ triggered, at least for the `T::Item` case, which means that a lifetime of '`static` is being assumed (and the error message indicates as such). Surprisingly, passing `-Z unpretty=mir` to the compiler results in the error message from #62849 appearing!

Even _with_ the error message, however, I'd be quite surprised if users intuit what the core issue is, in this case, that the compiler is unable to fill in the holes correctly, and that providing more type information resolves the issue. At minimum, an error code with a `--explain` should be added that indicates how to resolve the error in the interim.

**Comment by Mark-Simulacrum** [comment]
I believe this is all known information, but unmarking as triaged so that it is seen again by the async WG.

**Comment by tmandry** [comment]
I did some investigating, leaving notes here.

The reason the example [above](https://github.com/rust-lang/rust/issues/61949#issuecomment-551179083) does not emit the new error message is because the compile fails during type _collection_ and never makes it to type _checking_. This is a bug fixed by https://github.com/rust-lang/rust/pull/68884.

With the above patch applied, we now get the new error message that was introduced by https://github.com/rust-lang/rust/pull/62849. We _also_ get the more confusing error message about inferring appropriate lifetimes (this happens while checking the function body), but I'm not sure if we can avoid that without a lot of effort.

Agreed that the error message could be better. I think we should give this error a hint ("try spelling out the type instead of using `Self`"), along with an error code / longer description saying that this is a limitation of the compiler.

**Comment by tmandry** [comment]
Hello from triage.. downgrading from Focus since there is no forward-compat issue. We may keep this as OnDeck for now to track the diagnostics improvement.

**Comment by tmandry** [comment]
Opened #69276 to track the diagnostics improvement separately.

**Comment by Kimundi** [comment]
The situation here is still confusing, as the explanation for the error message, https://doc.rust-lang.org/nightly/error-index.html#E0760, talks about `async`, but the issue also arised with non-async code:

```rust
pub struct Foo<'a>(&'a ());

impl<'a> Foo<'a> {
    pub fn works<T>(self) -> impl FnOnce(T) -> Foo<'a> {
        move |_| todo!()
    }
    pub fn errors<T>(self) -> impl FnOnce(T) -> Self {
        move |_| todo!()
    }
}
```

**Comment by Xiretza** [comment]
#91403 was reverted, I think this should be reopened.

**Comment by Arnavion** [comment]
I just revisited the code that had prompted me to file this issue, and both it and the test case in the OP compile fine today. According to godbolt it didn't compile until 1.73 and started compiling since 1.74.0. Is this still an issue?

**Comment by tmandry** [comment]
Looks like this is now accepted after #115659 and we can close the issue. Thanks @Arnavion for commenting!
